# Homework Assignment #4 (Individual)
## Regression and SVM Predictive Models

### <p style="text-align: right;"> &#9989; Jack Holscher</p>
### <p style="text-align: right;"> &#9989; holscherj</p>

### Goal for this homework assignment
By now, you have learned a bit about predictive models. In this assignment, you will practice:

* Building predictive models using `statsmodels` and `sklearn`
* Evaluating your predictive models

**This assignment is due roughly two weeks from now at 11:59 pm on Friday, April 21st.** It should be uploaded into the "Homework Assignments" submission folder for Homework #4.  Submission instructions can be found at the end of the notebook.


---
## Part 0: Add to your Git repository to track your progress on your assignment (4 points)

For this assignment, you're going to add it to the `cmse202-s23-turnin` repository you created in class so that you can track your progress on the assignment and preserve the final version that you turn in. In order to do this you need to

**&#9989; Do the following**:

1. Navigate to your `cmse202-s23-turnin` repository and create a new directory called `hw-04`.
2. Move this notebook into that **new directory** in your repository, then **add it and commit it to your repository**.
1. Finally, to test that everything is working, "git push" the file so that it ends up in your GitHub repository.

**Important**: Double check you've added your Professor and your TA as collaborators to your "turnin" repository (you should have done this in the previous homework assignment).

**Also important**: Make sure that the version of this notebook that you are working on is the same one that you just added to your repository! If you are working on a different copy of the notebook, **none of your changes will be tracked**!

If everything went as intended, the file should now show up on your GitHub account in the "`cmse202-s23-turnin`" repository inside the `hw-04` directory that you just created.  Periodically, **you'll be asked to commit your changes to the repository and push them to the remote GitHub location**. Of course, you can always commit your changes more often than that, if you wish.  It can be good to get into a habit of committing your changes any time you make a significant modification, or when you stop working on the project for a bit.

&#9989; **Do this**: Before you move on, put the command that your instructor should run to clone your repository in the markdown cell below.

git clone https://github.com/holscherj/CMSE202-s23-turnin.git

# Building a Model to Predict if it Will Rain Tomorrow

In this dataset we have weather data for a number of Australian cities. We’re going to **build a model that uses data from today to predict if it will rain tomorrow.** To simplify our model, let’s just look at the weather for one city.

## Part 1. Working with the Data

The dataset that we’ll be using can be found here:
- `https://raw.githubusercontent.com/msu-cmse-courses/cmse202-S23-data/main/HW/HW4/aussie_weather_data.csv`

#### 1.1 (3 Points)
**Download the data set and read it in using Pandas.**

In [1]:
#Write your code here
import pandas as pd

weather_df = pd.read_csv("aussie_weather_data.csv")
weather_df

Year  Month  DayOfMonth       Location  MinTemp  MaxTemp  Rainfall  \
0      2009      1           1  NorfolkIsland     20.4     25.8       0.0   
1      2009      1           2  NorfolkIsland     20.9     26.7       0.2   
2      2009      1           3  NorfolkIsland     22.3     26.3       0.0   
3      2009      1           4  NorfolkIsland     21.6     22.2       1.2   
4      2009      1           5  NorfolkIsland     20.4     23.5       2.6   
...     ...    ...         ...            ...      ...      ...       ...   
39510  2017      6          20         Darwin     19.3     33.4       0.0   
39511  2017      6          21         Darwin     21.2     32.6       0.0   
39512  2017      6          22         Darwin     20.7     32.8       0.0   
39513  2017      6          23         Darwin     19.5     31.8       0.0   
39514  2017      6          24         Darwin     20.2     31.7       0.0   

       Evaporation  Sunshine  WindGustDir  ...  Humidity9am  Humidity3pm  \
0              6.0      12.4        112.5  ...         84.0         71.0   
1              8.0      10.3         90.0  ...         79.0         77.0   
2              3.2       2.0          0.0  ...         87.0         90.0   
3              2.8       0.0        337.5  ...         92.0         95.0   
4              2.2       2.9        337.5  ...         86.0         86.0   
...            ...       ...          ...  ...          ...          ...   
39510          6.0      11.0         22.5  ...         63.0         32.0   
39511          7.6       8.6          0.0  ...         56.0         28.0   
39512          5.6      11.0          0.0  ...         46.0         23.0   
39513          6.2      10.6        337.5  ...         62.0         58.0   
39514          5.6      10.7         22.5  ...         73.0         32.0   

       Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0           1014.5       1013.6       3.0       1.0     23.3     24.7   
1           1016.3       1015.5       2.0       5.0     25.0     25.1   
2           1014.6       1014.3       7.0       7.0     24.7     23.8   
3           1016.0       1015.3       8.0       8.0     22.1     21.2   
4           1015.3       1013.7       7.0       7.0     21.8     21.6   
...            ...          ...       ...       ...      ...      ...   
39510       1013.9       1010.5       0.0       1.0     24.5     32.3   
39511       1014.6       1011.2       7.0       0.0     24.8     32.0   
39512       1015.3       1011.8       0.0       0.0     24.8     32.1   
39513       1014.9       1010.7       1.0       1.0     24.8     29.2   
39514       1013.9       1009.7       6.0       5.0     25.4     31.0   

       RainToday  RainTomorrow  
0              0             0  
1              0             0  
2              0             1  
3              1             1  
4              1             0  
...          ...           ...  
39510          0             0  
39511          0             0  
39512          0             0  
39513          0             0  
39514          0             0  

[39515 rows x 25 columns]

#### 1.2 (3 Points)
This dataset has weather information for multiple cities in Australia. It’s quite a large dataset. In fact, it’s a little *too* large for our purposes; it will take a considerable amount of time to train a model on so much data. So we’ll just work with the data for a specific city.

The first thing we’ll do is see which cities we have data for. 
**Print out a list of the specific cities in this dataset.** (Note that this is not the same as printing the entire `Location` column.)

In [2]:
#Write your code here
weather_df['Location'].value_counts()

Darwin              3062
Perth               3025
Brisbane            2953
MelbourneAirport    2929
PerthAirport        2913
SydneyAirport       2870
Watsonia            2730
Mildura             2594
MountGambier        2465
NorfolkIsland       2464
Cairns              2444
Townsville          2419
WaggaWagga          2416
AliceSprings        2223
Nuriootpa           2008
Name: Location, dtype: int64

#### 1.3 (3 Points)
Select one of the cities (and **please** don't just pick the first one!). Cut down your dataframe so that it only contains data from your chosen city. You should also **look through the columns in this dataset and ensure that all of the data is ready for our analysis/model creation.**

In [12]:
#Write your code here
other = weather_df[weather_df["Location"] != "Brisbane"].index
weather_df.drop(other, inplace=True)

weather_df.isnull().sum()
weather_df

Year  Month  DayOfMonth  Location  MinTemp  MaxTemp  Rainfall  \
16003  2008      7           1  Brisbane     10.3     25.2       0.0   
16004  2008      7           2  Brisbane      8.1     22.9       0.0   
16005  2008      7           3  Brisbane      9.7     22.4       0.0   
16006  2008      7           4  Brisbane     11.8     20.0       0.8   
16007  2008      7           5  Brisbane     12.3     16.7       0.0   
...     ...    ...         ...       ...      ...      ...       ...   
18951  2017      6          21  Brisbane     12.2     23.4       0.0   
18952  2017      6          22  Brisbane     13.7     22.7       0.0   
18953  2017      6          23  Brisbane     10.3     23.7       0.0   
18954  2017      6          24  Brisbane     10.4     24.5       0.0   
18955  2017      6          25  Brisbane     11.0     24.2       0.0   

       Evaporation  Sunshine  WindGustDir  ...  Humidity9am  Humidity3pm  \
16003          1.4       9.5        157.5  ...         81.0         37.0   
16004          2.0       9.8        180.0  ...         41.0         30.0   
16005          5.8       9.4          0.0  ...         55.0         52.0   
16006          1.8       1.1        225.0  ...         76.0         53.0   
16007          2.0       0.3        270.0  ...         81.0         89.0   
...            ...       ...          ...  ...          ...          ...   
18951          4.0       9.2        202.5  ...         61.0         44.0   
18952          3.4       7.2        315.0  ...         62.0         55.0   
18953          3.2       9.4        180.0  ...         71.0         37.0   
18954          3.4       8.9        270.0  ...         75.0         33.0   
18955          2.2       9.8         22.5  ...         68.0         53.0   

       Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
16003       1019.6       1014.8       0.0       1.0     14.9     24.6   
16004       1018.8       1015.0       0.0       0.0     16.2     22.4   
16005       1021.4       1019.1       1.0       4.0     15.4     21.3   
16006       1023.5       1021.7       7.0       7.0     14.1     19.6   
16007       1027.3       1026.2       7.0       8.0     16.1     15.0   
...            ...          ...       ...       ...      ...      ...   
18951       1023.4       1021.3       1.0       4.0     17.2     23.3   
18952       1026.0       1022.4       2.0       3.0     17.8     20.5   
18953       1023.3       1018.4       1.0       3.0     15.2     23.1   
18954       1018.6       1015.4       7.0       5.0     14.3     24.0   
18955       1020.5       1017.3       6.0       3.0     15.9     22.6   

       RainToday  RainTomorrow  
16003          0             0  
16004          0             0  
16005          0             0  
16006          0             0  
16007          0             1  
...          ...           ...  
18951          0             0  
18952          0             0  
18953          0             0  
18954          0             0  
18955          0             0  

[2953 rows x 25 columns]

#### 1.4 (3 Points)
Finally, before we start creating our models, let's first split our data into training and testing datasets. **Keep in mind what we want our model to predict and how this dataset gives you the information you need for your model.**

In [4]:
#Write your code here
from sklearn.model_selection import train_test_split

features = ['Month', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation', 'Sunshine',
            'WindGustSpeed', 'WindSpeed9am', 'WindSpeed3pm',
            'Humidity9am', 'Humidity3pm', 'Pressure9am', 'Pressure3pm', 'Cloud9am',
            'Cloud3pm', 'Temp9am', 'Temp3pm', 'RainToday']

X = weather_df.loc[:, features]
y = weather_df.loc[:, 'RainTomorrow']

train_vectors, test_vectors, train_labels, test_labels = train_test_split(X, y, random_state=128, train_size=.75)

## 2.0 Building a Regression Model

#### 2.1 (3 Points)
We’ll start by creating a regression model. **Does it make sense for us to use linear regression or logistic regression? Explain your choice.**

*Since we are working with a classification model, determining whether or not it will rain tomorrow, logistic regression is the better choice since it works better in making decisions like this compared to linear.*

#### 2.2 (12 Points)
Create a regression model. Use the model summary to determine which parameters are most statistically significant in your model. If you get an error during at this stage it may be helpful to go back to part 1.3 and ensure that all of your data is appropriate for creating a model. 

In [5]:
#Write your code here
import statsmodels.api as sm

log_model = sm.Logit(train_labels, sm.add_constant(train_vectors))
result = log_model.fit()
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.314946
         Iterations 8
                           Logit Regression Results                           
Dep. Variable:           RainTomorrow   No. Observations:                 2214
Model:                          Logit   Df Residuals:                     2195
Method:                           MLE   Df Model:                           18
Date:                Wed, 19 Apr 2023   Pseudo R-squ.:                  0.3989
Time:                        11:53:40   Log-Likelihood:                -697.29
converged:                       True   LL-Null:                       -1160.1
Covariance Type:            nonrobust   LLR p-value:                5.315e-185
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
const           -40.2933     17.509     -2.301      0.021     -74.610      -5.977
Month             0.

/Users/jacksonholscher/opt/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


*The features that were most significant in the model, based on P-values, were Sunshine, WindGustSpeed, WindSpeed9am, Humidity3pm, Pressure9am, Pressure3pm, and Cloud3pm*

#### 2.3 (8 Points)
**Fit your test data and create/print a confusion matrix. We’ll use this to evaluate how well your predicitve model performs.**

In [6]:
#Write your code here
import numpy as np
from sklearn.metrics import confusion_matrix

values = result.predict(sm.add_constant(test_vectors))
predictions = np.where(values > 0.5, 1, 0)

mat = confusion_matrix(np.array(test_labels), predictions)
print(mat)

[[553  26]
 [ 81  79]]


/Users/jacksonholscher/opt/anaconda3/lib/python3.9/site-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


#### 2.4 (3 Points)

Look at/calculate the Precision and Recall of your model. **Which one is greater? Describe in plain language what that means about the performance of your model (I.e., the circumstances in which it does/doesn’t do well).**

In [7]:
from sklearn.metrics import precision_score, recall_score

p_score = precision_score(test_labels, predictions)
r_score = recall_score(test_labels, predictions)

print(f"Precision: {p_score}")
print(f"Recall: {r_score}")

correct = 0
total_samples = len(test_labels)

for i in range(len(test_labels)):
    if predictions[i] == np.array(test_labels)[i]:
        correct += 1

accuracy = correct / total_samples

print(f"Accuracy: {accuracy}")

Precision: 0.7523809523809524
Recall: 0.49375
Accuracy: 0.8552097428958051


*The precision of the model is greater than the recall. In machine learning, precision aims to measure the accuracy of positive predictions (true/false positives), while recall aims to measures how often the model can correctly identify the positive cases out of all actual positive cases. With a higher precision, this model will perform better when predicting positives, and could struggle with how many false negatives it predicts.*

## 3.0 Interlude: A Base Rate Frequency Model (12 Points)

If we didn't create a fancy model for predicting future rainfall, would that mean that we had *no way* of predicting if it would rain? Of course not. There's always *some* kind of model we can use (they just won't be very sophisticated). 

In the case of our problem--predicting whether it will rain tomorrow--what is the simplest model we could use? We'd need to look at the **base rate frequency**. If we know that, over the course of the year, it rains ~30% of the time, that would be our base rate frequency. If we were to guess that it would be sunny every day of the year, we'd be wrong ~30% of the time, *but we'd be right ~70% of the time.* 

Why is this important? Imagine we create a sophisticated model that accurately predicts rain/no rain ~60% of the time; without any other information, we might think, "That's not too bad, I guess." It certainly feels better than *no* information, right? Now think about the base rate frequency model, which achieves an accuracy of **70%** just by guessing the same thing every time. Suddenly, our sophisticated model seems like crap! It tells us we could be doing ~10% better than our fancy model by guessing. This is why we use the base rate frequency as a simple way to test our models.


**Calculate how frequently it rains in your city. If you were to guess that it *wasn't* going to rain every day, how frequently would you be right? (We'll use this as our baseline model.)**


In [8]:
#Write your code for calculating the frequency of rain here.
print(100 - weather_df["RainToday"].mean() * 100)

78.25939722316289


*In Brisbane, if I were to guess that it wasn't going to rain every day, I would be right 78.2593% of the time.*

---

### &#128721; STOP (3 Point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit the changes to your Git repository with a meaningful commit message.

---

## 4.0 SVM Model

We’ve seen how regression faired in predicting rain; now let’s give Support Vector Machines a shot. 

It’s not clear, *a priori*, which type of kernel (linear or RBF) would fair better. Therefore, we’ll need to test out both kernel types, as well as multiple values for hyperparameters ($C$ and $\gamma$). 

#### 4.1 (12 Points)
Run a grid search over both types of kernels and multiple values of $C$ and $\gamma$ (we recommend [0.01,0.1,1.0,10.0] for both). We *STRONGLY* recommend you set the parameter `n_jobs=-1` in `GridSearchCV`, which will distribute the computational load. Make sure to write down your best-fit hyperparameters. 

In [9]:
#Write your code here
from sklearn import svm
from sklearn.model_selection import GridSearchCV

vals = [0.01, 0.1, 1.0, 10.0]
params = {'C': vals, 'gamma': vals, 'kernel': ('linear', 'rbf')}
svc = svm.SVC()

search = GridSearchCV(svc, params, n_jobs=-1)
search.fit(train_vectors, train_labels)

print(search.best_params_)

{'C': 0.01, 'gamma': 0.01, 'kernel': 'linear'}


*The best-fit hyperparameters are 10 for the C value, 0.01 for the gamma, and a linear kernel.*

#### 4.2 (8 Points)
Using your best-fit parameters, classify your test data and print off the confusion matrix. Note/calculate the Precision and Recall for your SVM model, as well as the overall accuracy.

$\mathrm{Accuracy} = \frac{\mathrm{Accurate~Predictions}}{\mathrm{Total~Samples}}$.

In [10]:
#Write your code here
model = svm.SVC(C=10.0, gamma=0.01, kernel='linear')
model.fit(train_vectors, train_labels)

predictions = model.predict(test_vectors)
mat = confusion_matrix(test_labels, predictions)
print(f"Confusion matrix:\n {mat}\n")

p_score = precision_score(test_labels, predictions)
r_score = recall_score(test_labels, predictions)

print(f"Precision: {p_score}")
print(f"Recall: {r_score}")

correct = 0
total_samples = len(test_labels)

for i in range(len(test_labels)):
    if predictions[i] == np.array(test_labels)[i]:
        correct += 1

accuracy = correct / total_samples

print(f"Accuracy: {accuracy}")

Confusion matrix:
 [[559  20]
 [ 90  70]]

Precision: 0.7777777777777778
Recall: 0.4375
Accuracy: 0.851150202976996


## 5. Evaluation of Models

Here we’re going to investigate how well our predictive models actually work.


#### 5.1 (5 Points)

Compare the accuracy of both of your predictive models (regression and SVM) to the baseline model from part 3. Do your predictive models do better than the baseline model? Is their performance relative to the baseline model what *you* would expect? 


Regression: 
- Precision: 0.7777777777777778
- Recall: 0.5515151515151515
- Accuracy: 0.86468200270636

SVM: 
- Precision: 0.7647058823529411
- Recall: 0.5515151515151515
- Accuracy: 0.8619756427604871

As you can see from the results above, the accuracies are very similar, differing only by 0.00270635994. They had the same recall score and very similar precisions, with the regression model having a very slightly higher precision. Compared to the baseline model, which has an accuracy of 0.782593, the predictive models outperformed by a small but notable margin. I would expect the models to perform higher than the margin that they do. I imagine if I refined the models to only include the most effective features that I listed above in 2.2, they might have higher overall accuracies.

#### 5.2 (5 Points)

Let’s test the robustness of your model. Before you do anything, you should record the precision, recall, and accuracy of your two predictive models (if you haven’t already). Once you’ve done that, go all the way back to part 1.4 and change the random seed for `train_test_split`. Using this new data split, rerun all of your code from parts 2 and 4. For each new random seed, record the precision, recall, and accuracy for both models (and write them, as well as the random seed, in the cell below!). Do this for 3-5 different random seeds. 


- Random Seed 123 Listed Above


- Random Seed 124:
    * Regression
        - Precision: 0.7692307692307693
        - Recall: 0.5649717514124294
        - Accuracy: 0.8552097428958051

    * SVM
        - Precision: 0.8141592920353983
        - Recall: 0.519774011299435
        - Accuracy: 0.8565629228687416


- Random Seed 125:
    * Regression
        - Precision: 0.7681159420289855
        - Recall: 0.572972972972973
        - Accuracy: 0.8497970230040596

    * SVM
        - Precision: 0.8174603174603174
        - Recall: 0.5567567567567567
        - Accuracy: 0.857916102841678


- Random Seed 126:
    * Regression
        - Precision: 0.717948717948718
        - Recall: 0.5454545454545454
        - Accuracy: 0.8606224627875507

    * SVM
        - Precision: 0.76
        - Recall: 0.4935064935064935
        - Accuracy: 0.8619756427604871


- Random Seed 127:
    * Regression
        - Precision: 0.7884615384615384
        - Recall: 0.5061728395061729
        - Accuracy: 0.8619756427604871

    * SVM
        - Precision: 0.7745098039215687
        - Recall: 0.4876543209876543
        - Accuracy: 0.8565629228687416


- Random Seed 128:
    * Regression
        - Precision: 0.7523809523809524
        - Recall: 0.49375
        - Accuracy: 0.8552097428958051

    * SVM
        - Precision: 0.7777777777777778
        - Recall: 0.4375
        - Accuracy: 0.851150202976996

#### 5.3 (5 Points)

How much variation did you observe in the precision, recall, and accuracy in your models? After going through this exercise, how (if at all) does this change your response to part 5.1? 

*While there was some variance in the results between the different iterations, the differences in precision, recall, and accuracy are not very significant. Looking at precision and recall, it seemed that these values changed the most from random_state to random_state. Starting initially at ~0.777 and ~0.551 respectively, the precision jumped as high as ~0.817 for random_state = 125 and as low as ~0.717 for random_state = 126. Recall reached a low of ~0.437 at random_state = 128, and a high of ~0.572 at random_state = 125. Accuracy seemed to vary much less than the other two values. Initially coming in at around ~0.864, the highest value it hit, and it dipped down only to ~0.849 at its lowest. This doesn't really change my answer to 5.1. As you can see, the accuracy of the model still outperforms the baseline, in all cases, and I still think they would perform better if refined slightly. It's possible the lower recall values could pose an issue in some cases, but for this, I still think the high precision and accuracy make these models useful for predicting rain.*

#### 5.4 (5 Points)

Finally, compare the performance of your regression and SVM models. Which, if any, performed better? Esxplain.

In [11]:
regression_p = np.array([
    0.7692307692307693,
    0.7681159420289855,
    0.717948717948718,
    0.7884615384615384,
    0.7523809523809524,
    0.7777777777777778
])
regression_r = np.array([
    0.5649717514124294,
    0.572972972972973,
    0.5454545454545454,
    0.5061728395061729,
    0.49375,
    0.5515151515151515
])
regression_a = np.array([
    0.8552097428958051,
    0.8497970230040596,
    0.8606224627875507,
    0.8619756427604871,
    0.8552097428958051,
    0.86468200270636
])

svm_p = np.array([
    0.8141592920353983,
    0.8174603174603174,
    0.76,
    0.7745098039215687,
    0.7777777777777778,
    0.7647058823529411
])

svm_r = np.array([
    0.519774011299435,
    0.5567567567567567,
    0.4935064935064935,
    0.4876543209876543,
    0.4375,
    0.5515151515151515
])

svm_a = np.array([
    0.8565629228687416,
    0.857916102841678,
    0.8619756427604871,
    0.8565629228687416,
    0.851150202976996,
    0.8619756427604871
])

print(f"Regression, Avg Precision: {regression_p.mean()}, Recall: {regression_r.mean()}, Accuracy: {regression_a.mean()}")
print(f"SVM, Avg Precision: {svm_p.mean()}, Recall: {svm_r.mean()}, Accuracy: {svm_a.mean()}")

Regression, Avg Precision: 0.7623192829714568, Recall: 0.5391395434768786, Accuracy: 0.8579161028416779
SVM, Avg Precision: 0.7847688455913339, Recall: 0.5077844556775819, Accuracy: 0.8576905728461885


*Overall, based on the averages above from the six different random states, it appears that the models performed very similarly. The two models had a difference in accuracy that is likely negligible, so in my opinion, I think that the SVM is better, with a higher average precision. I think in general SVM is better than regression due to its ability to avoid overfitting data, as well as its adaptability to non-linear data through kernel functions.*

---

### &#128721; STOP (3 Point)
**Pause to commit your changes to your Git repository!**

Take a moment to save your notebook, commit the changes to your Git repository with a meaningful commit message.

---

### Congratulations, you're done!

&#169; Copyright 2022,  Department of Computational Mathematics, Science and Engineering at Michigan State University